libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp

```diff
@@ -0,0 +1,127 @@
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+  backend driver for airspeed from www.qio-tek.com a I2C ASP5033 sensor
+ */
+
+#include "AP_Airspeed_ASP5033.h"
+#include <GCS_MAVLink/GCS.h>
+#include <AP_Common/AP_Common.h>
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/I2CDevice.h>
+#include <AP_Math/AP_Math.h>
+
+extern const AP_HAL::HAL &hal;
+
+#define ASP5033_I2C_ADDR 0x6D
+#define Sys_config 0xA5
+#define CMD 0x30
+#define CMD_measure 0x0A
+#define Press_data 0x06
+#define Temp_data 0x09
+
+
+AP_Airspeed_ASP5033::AP_Airspeed_ASP5033(AP_Airspeed &_frontend, uint8_t _instance) :
+    AP_Airspeed_Backend(_frontend, _instance)
+{
+}
+
+// probe and initialise the sensor
+bool AP_Airspeed_ASP5033::init()
+{
+    uint8_t SysID;
+    _dev = hal.i2c_mgr->get_device(get_bus(), ASP5033_I2C_ADDR);
+    if (!_dev) {
+        return false;
+    GCS_SEND_TEXT(MAV_SEVERITY_INFO,"ASP5033: sensor no found\n");
+    }
+    _dev->get_semaphore()->take_blocking();
+    _dev->set_retries(2);
+    _dev->get_semaphore()->give();
+    _dev->read_registers(Sys_config, &SysID, 1); //Read Sensor_config
+    SysID = (SysID&0xFD);
+    _dev->write_register(Sys_config, SysID); //Set Sensor_config
+    _dev->register_periodic_callback(1000000UL/80U,
+                                    FUNCTOR_BIND_MEMBER(&AP_Airspeed_ASP5033::timer, void));
+    GCS_SEND_TEXT(MAV_SEVERITY_INFO,"ASP5033: sensor found\n");
+    return true;
+}
+
+// read the data from the sensor
+void AP_Airspeed_ASP5033::timer()
+{
+    uint8_t Sco;
+    uint8_t pressraw_bytes[3];
+    uint8_t tempraw_bytes[2];
+    _dev->write_register(CMD, CMD_measure); //once temperature conversion immediately followed by once sensor signal conversion
+    _dev->read_registers(CMD, &Sco, 1); //Judge whether Data collection is over
+    if ((Sco & 0x08) > 0){ 
+    _dev->read_registers(Press_data, pressraw_bytes, 3); 
+        //Read ADC output data of pressure
+        pressraw = (pressraw_bytes[0]<<24) + (pressraw_bytes[1]<<16) + (pressraw_bytes[2]<<8);
+        //Effective value converting 24 bit integer to 32 bit integer 
+        pressraw >>=15;
+        //Read ADC output data of temperature
+        _dev->read_registers(Temp_data, tempraw_bytes, 2);
+        tempraw = (tempraw_bytes[0]<<8) + tempraw_bytes[1];
+        // right-shift signed integer back to get correct measurement value
+        tempraw >>=8;
+    }
+        WITH_SEMAPHORE(sem); 
+        pressure_sum += pressraw;
```

Comment:
This should only be done inside the if, if there was new data. Otherwise this just accumulates stale readings, which would throw off averaging, and slow down the sensor response.

Reply:
done